gitrevise/utils.py

```diff
@@ -274,7 +274,14 @@ def cut_commit(commit: Commit) -> Commit:
 
     # Run an interactive git-reset to allow picking which pieces of the
     # patch should go into the first part.
-    index.git("reset", "--patch", final_tree.persist().hex(), "--", ".", stdout=None)
+    index.git(
+        "reset",
+        "--patch",
+        final_tree.persist().hex(),
+        "--",
+        *pathspecs if pathspecs else ".",
+        stdout=None,
+    )
```

Comment:
Does this work from a subdirectory (when pathspecs (or ".") are relative to the current directory)?

I'm actually using this daily, and happened to try it from a subdirectory now. It works if I use a repo-absolute path from within the subdirectory. But looking at the code, I'm surprised that that "." doesn't otherwise mean the subdirectory (not that I think it should).

Idea: You could make the tests run from a subdirectory to cover this aspect too.

Reply:
Is there an alternative to git reset?

This appears to be a quirk of git reset: Its pathspec arguments are not relative. Although (according to `man 7 gitattributes`) it is command defined whether they are relative, I think relative is the only sensible interpretation: It is what your shell completes, whereas a toplevel relative pathspec is denoted by for example a `:/` prefix. So if we want that, we would need to not use git reset, or work around it. It also explains how the dot becomes the toplevel directory.

Fixing it the obvious way (by prefixing the current directory) would necessitate parsing the pathspec, which looks a bit undesirable to have to deal with: Not just parsing wise (even parsing the short form of the magic character prefix requires knowing the full set of magic characters), but we need to understand it, to know whether the path is supposed to be toplevel relative or not.

Reply:
> This appears to be a quirk of git reset: Its pathspec arguments are not relative

No. According to my testing they are treated like any other pathspec (relative unless prefixed with `:/` or `/`).

If I run

```sh
mkdir a b
touch a/1 b/2
git add a b
cd a
git reset --patch -- 1
```

that works as expected (only includes `a/1`, not `b/2`).

Now if I cancel the last command and do

```sh
git reset --patch -- .
```

it will prompt to reset anything inside `a/`, so effectively the same.

> It works if I use a repo-absolute path from within the subdirectory.

This is surprising.
Also, relative paths don't work (they should).

This is a bug in `git-revise`.
By default it runs `git reset` with the Git worktree as working directory.

```python
        if cwd is None:
            cwd = getattr(self, "workdir", None)
```

Because of this, our `-- .` doesn't make a difference today (it is confusing though).
Once we fix this however, it will cause deviant behavior from `git reset`.
I think we should remove it.
I'm not sure why it was added in 74c59ad (Improve performance and UX for --cut, 2019-07-21), perhaps by accident.

Reply:
I agree. You found the right problem.

I had a go at running git from cwd (which indeed makes the pathspecs work from a subdirectory):
https://github.com/anordal/git-revise/commits/run-git-from-cwd

That getattr: I replaced it with `self.cwd`.
It might have been an approximation that was good enough until now:
The Repository constructor takes a cwd argument, so to respect that, it needs to run git from at least the right worktree.
Actually, it's unused, but it wasn't wrong to replace self.workdir with self.cwd AFAICS.

Reply:
I opted for a minimal fix here but I agree that it might be worth to change the default to keep the working directory, instead of using the work tree root.
However we need to be careful, your patch changes the behavior of our call to `git ls-files`. Not sure if for better or worse.